openedx/core/djangoapps/user_api/completion/views.py

```diff
@@ -0,0 +1,50 @@
+import csv
+
+from rest_framework.authentication import SessionAuthentication
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from rest_framework import permissions
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from .tasks import migrate_progress, OUTCOME_MIGRATED
+
+
+class MigrateProgressView(APIView):
+    """
+    Migrates user progress for a set of user pairs.
+    Only admins can use this.
+    """
+
+    authentication_classes = (JwtAuthentication, SessionAuthentication,)
+    permission_classes = (
+        permissions.IsAuthenticated,
+        permissions.IsAdminUser,
+    )
+
+    def post(self, request):
+        """
+        POST /api/user/v1/completion/migrate/
+
+        Migrate user progress.
+        """
+        csv_file = request.FILES['file']
+        recepient_address = request.POST.get('recepient_address')
+
+        reader = csv.DictReader(csv_file)
+
+        if not {'course', 'source_email', 'dest_email'}.issubset(set(reader.fieldnames)):
+            # Assert correct csv is uploaded
+            return Response(status=400)
+
+        # Extract list to be used in migration task
+        migrate_list = [
+            (row['course'], row['source_email'], row['dest_email']) for row in reader
+            if row.get('outcome') != OUTCOME_MIGRATED  # Ignore lines marked as migrated
+        ]
+
+        # Start background task to migrate progress for given users
+        migrate_progress.delay(
```

Comment:
Did you try this task running the workers externally? The default vagrant devstack behavior is to run celery on eager mode.

Reply:
@giovannicimolin, I use docker-based devstack. Tried to set `CELERY_ALWAYS_EAGER` to `False`, restarted lms, but received this error after sending POST to view:
```
Traceback:  

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _legacy_get_response
  249.             response = self._get_response(request)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py" in inner
  185.                     return func(*args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py" in wrapped_view
  58.         return view_func(*args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py" in view
  68.             return self.dispatch(request, *args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  489.             response = self.handle_exception(exc)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py" in handle_exception
  449.             self.raise_uncaught_exception(exc)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  486.             response = handler(request, *args, **kwargs)

File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/user_api/completion/views.py" in post
  57.             [recepient_address] if recepient_address else None

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py" in delay
  461.         return self.apply_async(args, kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py" in apply_async
  573.             **dict(self._get_exec_options(), **options)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/base.py" in send_task
  354.                 reply_to=reply_to or self.oid, **options

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/amqp.py" in publish_task
  310.             **kwargs

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/messaging.py" in publish
  172.                        routing_key, mandatory, immediate, exchange, declare)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in _ensured
  470.                                            interval_max)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in ensure_connection
  382.                         interval_start, interval_step, interval_max, callback)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/utils/__init__.py" in retry_over_time
  246.             return fun(*args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in connect
  250.         return self.connection

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in connection
  756.                 self._connection = self._establish_connection()

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in _establish_connection
  711.         conn = self.transport.establish_connection()

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/transport/pyamqp.py" in establish_connection
  116.         conn = self.Connection(**opts)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py" in __init__
  165.         self.transport = self.Transport(host, connect_timeout, ssl)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py" in Transport
  186.         return create_transport(host, connect_timeout, ssl)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/transport.py" in create_transport
  299.         return TCPTransport(host, connect_timeout)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/transport.py" in __init__
  95.             raise socket.error(last_err)

Exception Type: error at /api/user/v1/progress/migrate/
Exception Value: [Errno 111] Connection refused
```
I guess Celery worker should be started somehow.